security/pkg/nodeagent/cache/secretcache.go

```diff
@@ -256,60 +267,45 @@ func (sc *SecretManagerClient) GenerateSecret(resourceName string) (secret *secu
 		return ns, nil
 	}
 
-	if resourceName != RootCertReqResourceName {
-		if c := sc.cache.GetWorkload(); c != nil {
-			// cache hit
-			cacheLog.WithLabels("ttl", time.Until(c.ExpireTime)).Info("returned workload certificate from cache")
-			return c, nil
-		}
-		t0 := time.Now()
-		sc.generateMutex.Lock()
-		defer sc.generateMutex.Unlock()
-		if c := sc.cache.GetWorkload(); c != nil {
-			// Now that we got the lock, check again if there is a cached secret (from the caller holding the lock previously)
-			cacheLog.WithLabels("ttl", time.Until(c.ExpireTime)).Info("returned delayed workload certificate from cache")
-			return c, nil
-		}
-		if ts := time.Since(t0); ts > time.Second {
-			cacheLog.Warnf("slow generate secret lock: %v", ts)
-		}
-
-		ns, err := sc.generateSecret(resourceName)
-		if err != nil {
-			return nil, fmt.Errorf("failed to generate workload certificate: %v", err)
-		}
-
-		cacheLog.WithLabels("latency", time.Since(t0), "ttl", time.Until(ns.ExpireTime)).Info("generated new workload certificate")
-		sc.registerSecret(*ns)
+	ns = sc.getCachedSecret(resourceName)
+	if ns != nil {
 		return ns, nil
 	}
-	// If request is for root certificate,
-	// retry since rootCert may be empty until there is CSR response returned from CA.
-	rootCert, rootCertExpr := sc.cache.GetRoot()
-	if rootCert == nil {
-		wait := retryWaitDuration
-		retryNum := 0
-		for ; retryNum < maxRetryNum; retryNum++ {
-			time.Sleep(wait)
-			rootCert, rootCertExpr = sc.cache.GetRoot()
-			if rootCert != nil {
-				break
-			}
 
-			wait *= 2
-		}
+	t0 := time.Now()
+	sc.generateMutex.Lock()
+	defer sc.generateMutex.Unlock()
+
+	// Look at cache again before sending request to avoid overwhelming CA
+	ns = sc.getCachedSecret(resourceName)
+	if ns != nil {
+		return ns, nil
 	}
 
-	if rootCert == nil {
-		return nil, errors.New("failed to get root cert")
+	if ts := time.Since(t0); ts > time.Second {
+		cacheLog.Warnf("slow generate secret lock: %v", ts)
 	}
 
-	t := time.Now()
-	ns = &security.SecretItem{
-		ResourceName: resourceName,
-		RootCert:     rootCert,
-		ExpireTime:   rootCertExpr,
-		CreatedTime:  t,
+	// send request to CA to get new workload certificate
+	ns, err = sc.generateNewSecret(resourceName)
+	if err != nil {
+		return nil, fmt.Errorf("failed to generate workload certificate: %v", err)
+	}
+
+	// Store the new secret in the secretCache and trigger the periodic rotation for workload certificate
+	sc.registerSecret(*ns)
```

Comment:
The ROOTCA requests are not triggered upon expiry of any cert - they are triggered for two reasons (1) When there is a change in the workload certificate cert chain (2) when the proxy config certs changes. Also, note that when we "GENERATE" a new secret item, it is ALWAYS the Workload Certificate (we make no attempt to generate a ROOTCA). Also, we only GENERATE a new secret item if the workload certificate isn't already in the cache.

Reply:
Looking at `registerSecret`, I think this will always exit early with `resourceLog(item.ResourceName).Infof("skip scheduling certificate rotation, already scheduled")`. If it _doesn't_, then it will set the `ns` as the WorkloadCert, which is wrong since it just has the root cert field set.

Based on this, I think we would only want this for the workload cert, or I read the code wrong.

Reply:
Or at this point, the `ns` may be the full SecretItem? but from `getCachedSecret` it would just be a partial one?

Reply:
Yep. getCachedSecret() only returns the information needed for the request (ROOTCA or WORKLOAD_CERT). However, what we store in registerSecret comes from the CA directly and that contains all the information. 
I can understand that this is confusing - thats why in my earlier commit, I always filled up all fields of the secret Item, even in getCachedSecret(). What was ultimately passed to envoy depended on the resourceType in the secretItem.